Containers

jppaulo06 opened the pull request:
Com esse PR, colocamos containers no nosso projeto, orquestrados por Docker Compose.

O PR configura 3 containers:
- 1 container de Banco de Dados (postgres)
- 1 container para o servidor em PRODUÇÃO
- 1 container para o servidor em DESENVOLVIMENTO

O projeto ficará muito mais simples de ser executado, tanto em produção quanto em desenvolvimento.

Para produção:
```
docker compose --profile prod up
```

Para desenvolvimento:
```
docker compose --profile dev up
```
**OBS:** O ambiente de desenvolvimento **possui hot reload.**

Uma coisa a se atentar é que as configurações do banco de dados estão em .env.example, e então devem ser ajustadas em um arquivo .env.

Ainda mais, o  banco de dados possui um volume atrelado para deixá-lo persistente.

nahharris commented on README.md:
````diff
@@ -2,28 +2,48 @@
+docker-compose --profile dev up
+```
 
-Instale o python, crie um _venv_ com `python -m venv .venv`, ative com `source .venv/bin/activate` (repita isso sempre que abrir um novo _shell_ na pasta raíz do projeto). Com o _venv_ ativo, execute `pip install -r requirements.txt` para instalar as dependências e execute com `fastapi dev src/app.py` (sempre com o _venv_ ativo). 
````
Eu não removeria essa linha. Apesar de o docker resolver o problema da execução, sem a instalação das dependencias localmente o desenvolvedor não terá qualquer apoio do LSP.

"Mas eu posso conectar o VS Code ao container". Desse modo o desenvolvedor perde completamente o seu shell e qualquer utilitário quando estiver conectado ao container

jppaulo06 replied:
Entendo. Eu pensei em não remover, mas não vejo motivo de alguém executar o servidor sem o banco de dados configurado, daí removi.

É verdade, o ponto do LSP é importante. Vou colocar novamente, mas com algumas mudanças no texto.

nahharris commented on .env.example:
```diff
@@ -0,0 +1,3 @@
+POSTGRES_USER=postgres
+POSTGRES_DB=postgres
+POSTGRES_PASSWORD=change-me
```
Acho que poderiamos usar nomes diferentes para essas variáveis, por exemplo: `DBUSER`, `DBNAME`, `DBPASS`

jppaulo06 replied:
Se você prefere, vou mudar!

nahharris commented on README.md:
````diff
@@ -2,28 +2,48 @@
 
-- Python 3.12
+```bash
+docker-compose --profile dev up
````
Poderia mencionar também como encerrar a composição e que pode ser necessário executar um `docker compose build` quando fizer algumas modificações

jppaulo06 replied:
Ok!

nahharris commented:
Uma dúvida, como que o ambiente de desenvolvimento tem hot reload? A imagem copia os arquivos do projeto durante o build da imagem, certo? Teria que editar o código dentro do contêiner para que houvesse hot reload. 

Talvez para se ter hot reload fosse melhor montar a pasta atual em `/app`  dentro do conteiner e adicionar um hook pra rodar o pip sempre que o conteiner for inicializado

jppaulo06 commented:
> Uma dúvida, como que o ambiente de desenvolvimento tem hot reload? 

Quando rodado em desenvolvimento, o diretório `src` é montado no container como um volume. Além disso, o comando de rodar o servidor muda (`fastapi dev`).

> A imagem copia os arquivos do projeto durante o build da imagem, certo?

Sim,  com o `COPY . .`, copia-se todo o projeto para o /app do container.

> Teria que editar o código dentro do contêiner para que houvesse hot reload.

Está sendo editado com o volume :)

> Talvez para se ter hot reload fosse melhor montar a pasta atual em `/app` dentro do conteiner e adicionar um hook pra rodar o pip sempre que o conteiner for inicializado

A parte do pip é de fato relevante. É nisso que acho também importante a orientação de rodar `compose up` com a flag `--build` ou `docker compose build`. Só buildando a imagem novamente os pacotes são realmente atualizados no container.

jppaulo06 committed:
chore: Add Docker Image to project

Signed-off-by: Joao Paulo Pereira da Silva <jppaulo11@usp.br>

jppaulo06 committed:
feat: Add Docker Compose with Database

Add postgres database, put main server inside a container.

Add container to dev environment

Signed-off-by: Joao Paulo Pereira da Silva <jppaulo11@usp.br>

jppaulo06 committed:
docs: Update README to recommend containers

Now, Docker Compose is the default way to run the project, so show how
to run it correctly.

Signed-off-by: Joao Paulo Pereira da Silva <jppaulo11@usp.br>

nahharris commented:
Excelente trabalho João, PR aprovado e mergeado na `dev`
